### PR TITLE
polish up the ui per latest design

### PR DESCRIPTION
**Pull request by ben-zhang-at-salesforce**
Polish scanning screen UI based on Isabella's latest UI design
1. veiled surrounding, have a square transparent as scanning center. 
2. give webview consumer the capability to pass instructionText, successText. Default to empty and not showing
3. move the x button from top left to top right.
4. add the white check with green background as success indicator. 
5. draw blue box over the barcode area once acquired a success scan and show it for a second

### CODE REVIEW

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt [hunk 1]
```diff
@@ -198,10 +221,73 @@ class BarcodePluginActivity : AppCompatActivity() {
         eventBus.postSticky(FailedScanEvent(code, exception))
     }
 
+    private fun updateViewsForStartScan() {
+        if (barcodeScannerOptions?.instructionText!!.isNotEmpty()) {
```
Comment (jamie-houston) [inline platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt:225]:
```suggestion
        if (barcodeScannerOptions?.instructionText?.isNotEmpty() == true) {
```

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerOptions.kt [hunk 2]
```diff
@@ -13,7 +13,11 @@ import org.json.JSONObject
 import java.io.Serializable
 import java.lang.Exception
 
-class BarcodeScannerOptions(val barcodeTypes: List<BarcodeType> = listOf()) : Serializable {
+class BarcodeScannerOptions(
+    val barcodeTypes: List<BarcodeType> = listOf(),
+    val instructionText: String = "",
```
Comment (jamie-houston) [inline platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerOptions.kt:18]: I think you can just make this nullable and not add the "" everywhere.

File: platforms/android/barcodescanner/src/main/res/drawable/captured_area_indicator.xml [hunk 3]
```diff
@@ -1,3 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (jamie-houston) [inline platforms/android/barcodescanner/src/main/res/drawable/captured_area_indicator.xml:1]: nit: I think the copyright usually goes first. At least in this project...

File: platforms/android/barcodescanner/src/main/res/drawable/captured_area_indicator.xml [hunk 4]
```diff
@@ -6,10 +7,8 @@
   ~ For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
   ~
   -->
-
 <shape xmlns:android="http://schemas.android.com/apk/res/android"
     android:shape="rectangle" >
     <solid android:color="@color/transparent" />
-    <corners android:radius="@dimen/bounding_box_corner_radius" />
-    <stroke android:color="@color/white" android:width="@dimen/bounding_box_stroke_width" />
+    <stroke android:color="#148AED" android:width="4dp" />
```
Comment (jamie-houston) [inline platforms/android/barcodescanner/src/main/res/drawable/captured_area_indicator.xml:13]: nit: move to colors resource.  Not a big deal in a demo, but good practice... :shrug:

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt (hunk 1 above)
Comment (ben-zhang-at-salesforce) [reply platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt:225]: good idea

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerOptions.kt (hunk 2 above)
Comment (ben-zhang-at-salesforce) [reply platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerOptions.kt:18]: done

File: platforms/android/barcodescanner/src/main/res/drawable/captured_area_indicator.xml (hunk 3 above)
Comment (ben-zhang-at-salesforce) [reply platforms/android/barcodescanner/src/main/res/drawable/captured_area_indicator.xml:1]: android studio doesn't like the copyright comment go first  :(, show lots of red

image

File: platforms/android/barcodescanner/src/main/res/drawable/captured_area_indicator.xml (hunk 4 above)
Comment (ben-zhang-at-salesforce) [reply platforms/android/barcodescanner/src/main/res/drawable/captured_area_indicator.xml:13]: done

File: platforms/android/barcodescanner/build.gradle [hunk 5]
```diff
@@ -38,8 +39,6 @@ android {
 dependencies {
     def camerax_version = "1.0.0-beta03"
```
Comment (jamie-houston) [inline platforms/android/barcodescanner/build.gradle:40]: Should we update to beta10?

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt [hunk 6]
```diff
@@ -59,22 +70,30 @@ class BarcodePluginActivity : AppCompatActivity() {
...
         lifecycle.addObserver(viewFinder)
 
+        barcodeScannerOptions =
+            intent.extras?.getSerializable(OPTIONS_VALUE) as BarcodeScannerOptions?
         barcodeAnalyzer = BarcodeAnalyzer(
             this,
-            { qrCodes ->
-                firebase_ml_loading_indicator.visibility = View.GONE
+            { qrCodes, image ->
+                firebase_ml_loading_indicator.visibility = GONE
```
Comment (jamie-houston) [inline platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt:86]: We should either user `findViewById` or ktx extensions (like this line)

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt [hunk 7]
```diff
@@ -198,10 +221,73 @@ class BarcodePluginActivity : AppCompatActivity() {
...
+            statusText.text = barcodeScannerOptions?.instructionText
+        }
+        scanningIndicator.visibility = VISIBLE
+        scanSuccessIndicator.visibility = GONE
+        focusBox.isSelected = false
+        frozenFrameWrapper.visibility = GONE
+    }
+
+    private fun updateViewsForScanSuccess(barcode: FirebaseVisionBarcode, image: ImageProxy) {
+        if (barcodeScannerOptions?.successText!!.isNotEmpty()) {
```
Comment (jamie-houston) [inline platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt:236]: remove `!!`.  
```suggestion
        if (barcodeScannerOptions?.successText?.isNotEmpty() == true) {
        statusText.text = barcodeScannerOptions?.successText ?: "" // or something here...?
```

Comment (jamie-houston) [reply platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt:236]: Or you could pull this out into a helper that takes `barcodeScannerOptions?.instructionsText` or `barcodeScannerOptions?.successText` and does it... 🤷

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerOptions.kt [hunk 8]
```diff
@@ -9,29 +9,18 @@
...
+ * @Serializable to make json Serializable for kotlin
+ * extends Serializable to make it could be passed as activity bundle
+ * without extra code
+ */
+@Serializable
+class BarcodeScannerOptions(
+    val barcodeTypes: List<BarcodeType> = BarcodeType.values().asList(),
+    val instructionText: String? = null,
+    val successText: String? = null
+) : java.io.Serializable
```
Comment (jamie-houston) [inline platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerOptions.kt:26]: Does this need to inherit `java.io.Serializable`?

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerPlugin.kt [hunk 9]
```diff
@@ -35,7 +34,15 @@ class BarcodeScannerPlugin(private val context: Context) : Plugin, BarcodeScanne
         options: BarcodeScannerOptions?,
         callback: (barcode: BarcodeScannerResult?, failure: BarcodeScannerFailure?) -> Unit
     ) {
-        barcodeOptions = options ?: BarcodeScannerOptions(listOf())
+        // make sure have options and support all barcode types if not specified by client
+        barcodeOptions = options ?: BarcodeScannerOptions()
+        if (barcodeOptions?.barcodeTypes.isEmpty()) {
+            barcodeOptions = BarcodeScannerOptions(
```
Comment (jamie-houston) [inline platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerPlugin.kt:40]: I think this logic should be in `BarcodeScannerOptions`, so the default for BarcodeScannerOptions is all types.  I thought that was the way it was originally.

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerPlugin.kt [hunk 10]
```diff
@@ -35,7 +34,15 @@ class BarcodeScannerPlugin(private val context: Context) : Plugin, BarcodeScanne
...
+                BarcodeType.values().asList(),
```
Comment (jamie-houston) [inline platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerPlugin.kt:41]: I think this logic should be in `BarcodeScannerOptions`, so the default for BarcodeScannerOptions is all types.  I thought that was the way it was originally.

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerPlugin.kt [hunk 11]
```diff
@@ -35,7 +34,15 @@ class BarcodeScannerPlugin(private val context: Context) : Plugin, BarcodeScanne
...
+                barcodeOptions.instructionText,
+                barcodeOptions.successText)
```
Comment (jamie-houston) [inline platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerPlugin.kt:43]: I feel even more stronger after this logic... it creates the object, then creates a new object with the existing properties... like cloning.  I think it should just use `BarcodeType.values()` as the default if the list passed in is null or empty.

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeType.kt [hunk 12]
```diff
@@ -10,19 +10,45 @@
 package com.salesforce.barcodescannerplugin
 
 import com.google.firebase.ml.vision.barcode.FirebaseVisionBarcode
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
 
+@Serializable
 enum class BarcodeType {
+    @SerialName("code128")
```
Comment (jamie-houston) [inline platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeType.kt:18]: Could we just declare the BarcodeType serializer to toLowerCase() the name?  Or would that be more confusing?  I worry about having to hard code that for every value.

File: platforms/android/build.gradle [hunk 13]
```diff
@@ -25,10 +26,16 @@ allprojects {
         mavenLocal()
```
Comment (jamie-houston) [inline platforms/android/build.gradle:26]: We should remove mavenLocal from the repos

**Review by jamie-houston — APPROVED** [review]
Left some feedback to review before merging.

File: platforms/android/barcodescanner/build.gradle (hunk 5 above)
Comment (ben-zhang-at-salesforce) [reply platforms/android/barcodescanner/build.gradle:40]: upgraded to beta10 for camerax and alpha17 for camera-view.  minor change to make it work

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt (hunk 6 above)
Comment (ben-zhang-at-salesforce) [reply platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt:86]: make sense, done

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerOptions.kt (hunk 8 above)
Comment (ben-zhang-at-salesforce) [reply platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerOptions.kt:26]: nope, @Serializable only make kotinx happy, not activity bundle in the intent

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt (hunk 7 above)
Comment (ben-zhang-at-salesforce) [reply platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodePluginActivity.kt:236]: moved into a separate function, looks cleaner.

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerPlugin.kt (hunk 10 above)
Comment (ben-zhang-at-salesforce) [reply platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerPlugin.kt:41]: yes, originally it's in barcode options. 
after change it from json encodable to kotlinx serializable, there's not a good way to default the list to all if empty.
meanwhile I kind of think it's the ...plugin.kt's responsibility to decide how to the handle the empty list as input

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerPlugin.kt (hunk 11 above)
Comment (ben-zhang-at-salesforce) [reply platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerPlugin.kt:43]: fixed by default to empty list, then in init {} add all them in if empty to make sure it works event for "{"barcodeTypes": [], ...}"

Comment (ben-zhang-at-salesforce) [reply platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeScannerPlugin.kt:43]: fixed.

File: platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeType.kt (hunk 12 above)
Comment (ben-zhang-at-salesforce) [reply platforms/android/barcodescanner/src/main/java/com/salesforce/barcodescannerplugin/BarcodeType.kt:18]: @SerialName seems to only accept compile-time constant :(  let me know if you have better idea. 
image

File: platforms/android/build.gradle (hunk 13 above)
Comment (ben-zhang-at-salesforce) [reply platforms/android/build.gradle]: done

**Review by jamie-houston — APPROVED** [review]
looks good!

**Pull request merged**